Modules/_datetimemodule.c

```diff
@@ -416,6 +416,9 @@ iso_week1_monday(int year)
 static int
 iso_to_ymd(const int iso_year, const int iso_week, const int iso_day,
            int *year, int *month, int *day) {
+    if (iso_year <= 0) {
+        return -2;
```

Comment:
-2 means a specific kind of error (wrong week).

I think that the year value should be checked before calling `iso_to_ymd()`. It is already checked in one case, but not in other.

Please add tests.

Reply:
Based on the back trace, the year value is taken from the string at parse_iso format_date() (the caller of the iso_to_umd()). Do you recommend check it in parse_iso format_date() and return -1 (-1:  Failed to parse date component)? 
```
#10 0x00007ffff7e3d7d6 in iso_to_ymd (iso_year=0, iso_week=25, iso_day=1, year=year@entry=0x7fffffffd71c, month=month@entry=0x7fffffffd720, day=day@entry=0x7fffffffd724)
    at ./Modules/_datetimemodule.c:440
#11 0x00007ffff7e3d998 in parse_isoformat_date (dtstr=dtstr@entry=0x7ffff74dd328 "0000W25", len=len@entry=7, year=year@entry=0x7fffffffd71c, month=month@entry=0x7fffffffd720, 
    day=day@entry=0x7fffffffd724) at ./Modules/_datetimemodule.c:803
#12 0x00007ffff7e479c0 in datetime_fromisoformat (cls=0x7ffff7e58000 <PyDateTime_DateTimeType>, dtstr=0x7ffff74dd300) at ./Modules/_datetimemodule.c:5506
```
Do you mean add testcase to 'def test_fromisoformat_date_examples(self)' ?

Reply:
Currently, there are two calls of `iso_to_umd()`. In one case the year value is validated before calling `iso_to_umd()`, in other case it is not validated. We should either move the validation into `iso_to_umd()` (and introduce a special error code for this), or always validate the year value before calling `iso_to_umd()`.

Reply:
Thank you for comments. The fix is updated accordingly. 
Also I added the test case to Lib/test/datetimetester.py.